src/rust/engine/process_execution/src/remote_cache.rs

```diff
@@ -450,47 +450,70 @@ impl crate::CommandRunner for CommandRunner {
     )
     .await?;
 
-    // Check the remote Action Cache to see if this request was already computed.
-    // If so, return immediately with the result.
-    if self.cache_read {
-      let response = with_workunit(
-        context.workunit_store.clone(),
-        "check_action_cache".to_owned(),
-        WorkunitMetadata::with_level(Level::Debug),
-        crate::remote::check_action_cache(
-          action_digest,
-          &self.metadata,
-          self.platform,
-          &context,
-          self.action_cache_client.clone(),
-          self.store.clone(),
-          self.eager_fetch,
-        ),
-        |_, md| md,
-      )
-      .await;
-      match response {
-        Ok(cached_response_opt) => {
-          log::debug!(
-            "remote cache response: digest={:?}: {:?}",
+    let mut local_execution_future = self.underlying.run(req, context.clone());
+
+    let result = if self.cache_read {
+      // A future to read from the cache and log the results accordingly.
+      let cache_read_future = async {
+        let response = with_workunit(
+          context.workunit_store.clone(),
+          "check_action_cache".to_owned(),
+          WorkunitMetadata::with_level(Level::Debug),
+          crate::remote::check_action_cache(
             action_digest,
+            &self.metadata,
+            self.platform,
+            &context,
+            self.action_cache_client.clone(),
+            self.store.clone(),
+            self.eager_fetch,
+          ),
+          |_, md| md,
+        )
+        .await;
+        match response {
+          Ok(cached_response_opt) => {
+            log::debug!(
+              "remote cache response: digest={:?}: {:?}",
+              action_digest,
+              cached_response_opt
+            );
+            if cached_response_opt.is_some() {
+              context
+                .workunit_store
+                .increment_counter(Metric::SpeculationRemoteCacheCompletedFirst, 1);
+            }
```

Comment:
@tdyas @stuhood my thinking for metrics: we want to answer:

- When did remote caching save us from needing to use local execution?
- When was the remote caching roundtrip too slow and local execution won out?

So, I added `speculation_local_execution_completed_first` and `speculation_remote_cache_completed_first`.

Notably, I am not incrementing any counters for when this remote caching arm wins out in speculation, but there was a cache hit. I reason that there's nothing special about this case worth calling out - the cache was empty, but checking the remote cache wasn't unusually slow.

How does that sound?

Reply:
I don't have a preference on the naming, but @tdyas might.

But this metric increment should move. We can't know on this line that we won the race: we can only know that inside the relevant block of the `select` statement.